### PR TITLE
Actions: Allow actions to be called on the server

### PR DESCRIPTION
**Pull request by bholmesdev**
## Changes

Allow actions to be called on the server.

```astro
---
// src/pages/blog/[postId].astro
import { actions } from 'astro:actions';

await actions.like({ postId: Astro.params.postId });
---
```

- Ensure API context exists by wrapping the `next()` call from middleware with a context provider.
- Add an SSR code path to the `actions.mjs` template. This imports the `getAction()` utility to retrieve the appropriate action by path name.

## Testing

Add e2e test for calling the like action from astro frontmatter.

## Docs

TODO: RFC update

### CODE REVIEW

File: .changeset/eighty-taxis-wait.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+---
+"astro": minor
+---
+
+Allow actions to be called on the server. This allows you to call actions as utility functions in your Astro frontmatter or server-side components.
```
Comment (florian-lefebvre) [inline .changeset/eighty-taxis-wait.md:5]: Does it also allow using them in endpoints?

File: packages/astro/src/actions/runtime/middleware.ts
Comment (florian-lefebvre) [inline packages/astro/src/actions/runtime/middleware.ts:1]: Does it mean `getApiContext` could technically be called outside of actions?

File: packages/astro/src/core/render-context.ts [hunk 2]
```diff
@@ -8,7 +8,7 @@ import type {
 	RouteData,
 	SSRResult,
 } from '../@types/astro.js';
-import type { ActionAPIContext } from '../actions/runtime/store.js';
+import { type ActionAPIContext } from '../actions/runtime/store.js';
```
Comment (florian-lefebvre) [inline packages/astro/src/core/render-context.ts:11]: Why this change?

File: .changeset/eighty-taxis-wait.md (hunk 1 above)
Comment (bholmesdev) [reply .changeset/eighty-taxis-wait.md:5]: Yup! Just didn’t expect that use case as much. I’ll update the changeset

File: packages/astro/src/actions/runtime/middleware.ts
Comment (bholmesdev) [reply packages/astro/src/actions/runtime/middleware.ts:1]: Technically yes… but I’m not a huge fan that. Relying on local storage can be brittle for custom functions. I’m considering moving the `APIContext` back to the second handler parameter, and making this function internal only. Thoughts?

File: packages/astro/src/core/render-context.ts (hunk 2 above)
Comment (bholmesdev) [reply packages/astro/src/core/render-context.ts:11]: My editor being dumb

File: packages/astro/src/actions/runtime/middleware.ts
Comment (florian-lefebvre) [reply packages/astro/src/actions/runtime/middleware.ts:1]: I think it would be better yeah, I wonder if it should be the first or second parameter, I guess it depends on what we expect people to use more

Comment (bholmesdev) [reply packages/astro/src/actions/runtime/middleware.ts:1]: Erm, well if people are reaching for context more than the input we designed something by wrong 😅 But yeah open to what feels natural to you

Comment (florian-lefebvre) [reply packages/astro/src/actions/runtime/middleware.ts:1]: I think `(input, context)` makes sense 👍

File: .changeset/eighty-taxis-wait.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+---
+"astro": minor
```
Comment (ematipico) [inline .changeset/eighty-taxis-wait.md:2]: Since actions are experimental, I think we can flag this as `patch`. But I leave the decision to you

File: packages/astro/src/actions/runtime/middleware.ts [hunk 4]
```diff
@@ -47,15 +46,15 @@ export const onRequest = defineMiddleware(async (context, next) => {
 		},
 	};
 	Object.defineProperty(locals, '_actionsInternal', { writable: false, value: actionsInternal });
-	return next();
+	return ApiContextStorage.run(context, () => next());
```
Comment (ematipico) [inline packages/astro/src/actions/runtime/middleware.ts:49]: Just a word of warning... when users use the rewrites APIs inside the middleware, under the hood, we create a new `APIContext`; there's a chance that this `context` could be outdated in case users decide to use the rewrite APIs:

```js
next("/new-url") // new API, it will generate a new context with a new `Request`, and the URL is `/new-url`
```

**Review by ematipico — APPROVED** [review]

File: packages/astro/src/actions/runtime/middleware.ts (hunk 4 above)
Comment (bholmesdev) [reply packages/astro/src/actions/runtime/middleware.ts:49]: Ah, that's a good catch! I _think_ applying the `APIContext` to this case would resolve our issue, since it only gets called when a rewrite is detected: https://github.com/withastro/astro/pull/11088/files#diff-a07f25105bdef494ce45893f063230e218d44b36cd7db7c9c8d6447cb03db136R19

I'll need a unit test to confirm

Comment (ematipico) [reply packages/astro/src/actions/runtime/middleware.ts]: It's also true that having a rewrite inside an action feels weird, not sure there's a valid case for that. I could create a stackblitz after the release and see how that works

Comment (bholmesdev) [reply packages/astro/src/actions/runtime/middleware.ts]: Added a test. It works!
